gloo/transport/uv/pair.cc

```diff
@@ -0,0 +1,530 @@
+    uint64_t tag,
+    size_t offset,
+    size_t nbytes) {
+  auto buf = static_cast<UnboundBuffer*>(tbuf)->getWeakNonOwningPtr();
+
+  GLOO_ENFORCE_GE(offset, 0);
+  GLOO_ENFORCE_GE(nbytes, 0);
+  if (nbytes > 0) {
+    GLOO_ENFORCE_LE(offset, tbuf->size);
+    GLOO_ENFORCE_LE(nbytes, tbuf->size - offset);
+  }
+
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  // TODO(check if the pair is in the right state)
+
+  // Execute this send if there is a remote pending receive.
+  ContextMutator mutator(*context_, tag, rank_);
+  if (mutator.getRemotePendingRecv() > 0) {
+    // We keep a count of remote pending send and receive operations.
+    // In this code path the remote side hasn't seen a notification
+    // for this send operation yet so we need to take special care
+    // their count is updated regardless.
+    sendNotifySendReady(tag, nbytes);
+    sendUnboundBuffer(tag, NonOwningPtr<UnboundBuffer>(buf), offset, nbytes);
+    mutator.updateRemotePendingRecv(-1);
+    return;
+  }
+
+  // Notify peer of this pending send.
+  localPendingSend_[tag].emplace_back(std::move(buf), offset, nbytes);
+  sendNotifySendReady(tag, nbytes);
+}
+
+// Receive into the specified buffer from the remote side of pair.
+void Pair::recv(
+    transport::UnboundBuffer* tbuf,
+    uint64_t tag,
+    size_t offset,
+    size_t nbytes) {
+  auto buf = static_cast<UnboundBuffer*>(tbuf)->getWeakNonOwningPtr();
+
+  GLOO_ENFORCE_GE(offset, 0);
+  GLOO_ENFORCE_GE(nbytes, 0);
+  if (nbytes > 0) {
+    GLOO_ENFORCE_LE(offset, tbuf->size);
+    GLOO_ENFORCE_LE(nbytes, tbuf->size - offset);
+  }
+
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  // Notify peer of this pending recv.
+  localPendingRecv_[tag].emplace_back(std::move(buf), offset, nbytes);
+  sendNotifyRecvReady(tag, nbytes);
+
+  // Decrease balance of remote pending sends.
+  // This results in a negative count if executed before
+  // receiving the send notification from our peer.
+  ContextMutator mutator(*context_, tag, rank_);
+  mutator.updateRemotePendingSend(-1);
+}
+
+bool Pair::tryRecv(
+    transport::UnboundBuffer* tbuf,
+    uint64_t tag,
+    size_t offset,
+    size_t nbytes) {
+  auto buf = static_cast<UnboundBuffer*>(tbuf)->getWeakNonOwningPtr();
+
+  GLOO_ENFORCE_GE(offset, 0);
+  GLOO_ENFORCE_LT(offset, tbuf->size);
+  GLOO_ENFORCE_GT(nbytes, 0);
+  GLOO_ENFORCE_LE(nbytes, tbuf->size - offset);
+
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  // todo check if state is right...
+  //    throwIfException();
+
+  // Return early if there is no remote pending send.
+  ContextMutator mutator(*context_, tag, rank_);
+  if (mutator.getRemotePendingSend() <= 0) {
+    return false;
+  }
+
+  // Notify peer of this pending recv.
+  localPendingRecv_[tag].emplace_back(std::move(buf), offset, nbytes);
+  sendNotifyRecvReady(tag, nbytes);
+
+  // Decrease balance of remote pending sends.
+  mutator.updateRemotePendingSend(-1);
+  return true;
+}
+
+void Pair::close() {
+  // abort();
```

Comment:
Why do we need this comment here?

Reply:
Good catch. Need to fix and update this.